clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -127,7 +127,9 @@ public RecordAccumulator(LogContext logContext,
         this.closed = false;
         this.flushesInProgress = new AtomicInteger(0);
         this.appendsInProgress = new AtomicInteger(0);
-        this.batchSize = batchSize;
+        // As per Kafka producer configuration documentation batch.size may be set to 0 to explicitly disable
+        // batching which in practice actually means using a batch size of 1.
+        this.batchSize = Math.max(1, batchSize);
```

Comment:
Would it make sense to do this in `KafkaProducer` when the config is retrieved?

Reply:
Is there a reason why we'd prefer that? I only see disadvantages in sanitizing this value there:

1. In `KafkaProducer` the only other use of this config is in `BufferPool` which has not changed for a long while. The breaking behavior stems from recent changes that are localized to the `RecordAccumulator` and the `BuiltinPartitioner`. It does seem like `BufferPool` will work the same with buffer size either 0 or 1, but still, **if we focus this fix on the components where the bug was introduced we avoid causing further unintentional changes to previous behavior**.
2. If the change is made in `RecordAccumulator` **we can use the partition change callbacks to test for this bug**. To detect this bug with a test against `KafkaProducer` we'll need to execute `producer.send(record)` in a separate thread under a timeout, which isn't as nice.

Reply:
Whether code changed or not is not the defining point on how to design things. The comment you added to the code seems generally applicable and not specific to `RecordAccumulator`. If batch size of 0 really means batch size of 1, we should do that consistently across the components. Or else we may have regressions in other components the same way we had one here.

Reply:
Good point, @ijuma . Besides, I'm thinking we should have an integration test from producer to test under batch.size = 0, we can still successfully send records. This way we can make sure `BufferPool` and `RecordAccumulator` both work as expected.

Reply:
Moved to KafkaProducer and added an integration test. PTAL